# Question 6 Solutions using Data from Q3

In [1]:
# Import packages and modules

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# Import ast to use literal_eval
import ast

In [2]:
# Read the dataset

df = pd.read_excel('pokemon.xlsx')
df.head()

id  num        name                                               img  \
0   1    1   Bulbasaur  http://www.serebii.net/pokemongo/pokemon/001.png   
1   2    2     Ivysaur  http://www.serebii.net/pokemongo/pokemon/002.png   
2   3    3    Venusaur  http://www.serebii.net/pokemongo/pokemon/003.png   
3   4    4  Charmander  http://www.serebii.net/pokemongo/pokemon/004.png   
4   5    5  Charmeleon  http://www.serebii.net/pokemongo/pokemon/005.png   

                  type  height  weight             candy  candy_count  \
0  ['Grass', 'Poison']    0.71     6.9   Bulbasaur Candy         25.0   
1  ['Grass', 'Poison']    0.99    13.0   Bulbasaur Candy        100.0   
2  ['Grass', 'Poison']    2.01   100.0   Bulbasaur Candy          NaN   
3             ['Fire']    0.61     8.5  Charmander Candy         25.0   
4             ['Fire']    1.09    19.0  Charmander Candy        100.0   

           egg  spawn_chance  avg_spawns spawn_time multipliers  \
0            2         0.690        69.0      20:00      [1.58]   
1  Not in Eggs         0.042         4.2      07:00  [1.2, 1.6]   
2  Not in Eggs         0.017         1.7      11:30         NaN   
3            2         0.253        25.3      08:45      [1.65]   
4  Not in Eggs         0.012         1.2      19:00      [1.79]   

                             weaknesses  \
0  ['Fire', 'Ice', 'Flying', 'Psychic']   
1  ['Fire', 'Ice', 'Flying', 'Psychic']   
2  ['Fire', 'Ice', 'Flying', 'Psychic']   
3           ['Water', 'Ground', 'Rock']   
4           ['Water', 'Ground', 'Rock']   

                                      next_evolution  \
0  [{'num': '002', 'name': 'Ivysaur'}, {'num': '0...   
1               [{'num': '003', 'name': 'Venusaur'}]   
2                                                NaN   
3  [{'num': '005', 'name': 'Charmeleon'}, {'num':...   
4              [{'num': '006', 'name': 'Charizard'}]   

                                      prev_evolution  
0                                                NaN  
1              [{'num': '001', 'name': 'Bulbasaur'}]  
2  [{'num': '001', 'name': 'Bulbasaur'}, {'num': ...  
3                                                NaN  
4             [{'num': '004', 'name': 'Charmander'}]

#### Get all Pokemons whose spawn rate is less than 5%

In [3]:
pokemon_with_spawn_rate_less_than_5percent = list(df[df["spawn_chance"] < 0.05]["name"])

In [4]:
pokemon_with_spawn_rate_less_than_5percent

['Ivysaur',
 'Venusaur',
 'Charmeleon',
 'Charizard',
 'Wartortle',
 'Blastoise',
 'Butterfree',
 'Raichu',
 'Sandslash',
 'Nidoqueen',
 'Nidoking',
 'Clefable',
 'Ninetales',
 'Wigglytuff',
 'Vileplume',
 'Dugtrio',
 'Persian',
 'Primeape',
 'Arcanine',
 'Poliwrath',
 'Kadabra',
 'Alakazam',
 'Machoke',
 'Machamp',
 'Victreebel',
 'Golem',
 'Rapidash',
 'Slowbro',
 'Magneton',
 "Farfetch'd",
 'Dewgong',
 'Muk',
 'Cloyster',
 'Gengar',
 'Electrode',
 'Exeggutor',
 'Marowak',
 'Hitmonlee',
 'Hitmonchan',
 'Lickitung',
 'Weezing',
 'Rhydon',
 'Chansey',
 'Kangaskhan',
 'Seadra',
 'Starmie',
 'Mr. Mime',
 'Gyarados',
 'Lapras',
 'Ditto',
 'Vaporeon',
 'Jolteon',
 'Flareon',
 'Porygon',
 'Omastar',
 'Kabutops',
 'Aerodactyl',
 'Snorlax',
 'Articuno',
 'Zapdos',
 'Moltres',
 'Dragonair',
 'Dragonite',
 'Mewtwo',
 'Mew']

In [5]:
len(pokemon_with_spawn_rate_less_than_5percent)

65

Total there 65 pokemons which have spawn rate of less than 5 percent.

#### Get all Pokemons that have less than 4 weaknesses

In [6]:
# Using strips and replace to convert string to list to list for analysis

len(df["weaknesses"][0].strip("\'][ \'").replace("\'", "").replace(" ", "").split(","))

4

In [7]:
# Defining a function to get the length of the embedded list

def get_weakness_count(s: str) -> int:
    """
    Returns the number of weaknesses for a pokemon
    """
    
    return len(s.strip("\'][ \'").replace("\'", "").replace(" ", "").split(","))

In [8]:
pokemon_with_weakness_less_than_4 = list(df[df["weaknesses"].apply(lambda x: get_weakness_count(x)) < 4]["name"])

In [9]:
pokemon_with_weakness_less_than_4

['Charmander',
 'Charmeleon',
 'Charizard',
 'Squirtle',
 'Wartortle',
 'Blastoise',
 'Caterpie',
 'Metapod',
 'Pidgey',
 'Pidgeotto',
 'Pidgeot',
 'Rattata',
 'Raticate',
 'Spearow',
 'Fearow',
 'Ekans',
 'Arbok',
 'Pikachu',
 'Raichu',
 'Sandshrew',
 'Sandslash',
 'Nidoran ♀ (Female)',
 'Nidorina',
 'Nidoran ♂ (Male)',
 'Nidorino',
 'Clefairy',
 'Clefable',
 'Vulpix',
 'Ninetales',
 'Jigglypuff',
 'Wigglytuff',
 'Diglett',
 'Dugtrio',
 'Meowth',
 'Persian',
 'Psyduck',
 'Golduck',
 'Mankey',
 'Primeape',
 'Growlithe',
 'Arcanine',
 'Poliwag',
 'Poliwhirl',
 'Abra',
 'Kadabra',
 'Alakazam',
 'Machop',
 'Machoke',
 'Machamp',
 'Tentacool',
 'Tentacruel',
 'Ponyta',
 'Rapidash',
 'Magnemite',
 'Magneton',
 "Farfetch'd",
 'Doduo',
 'Dodrio',
 'Seel',
 'Grimer',
 'Muk',
 'Shellder',
 'Drowzee',
 'Hypno',
 'Krabby',
 'Kingler',
 'Voltorb',
 'Electrode',
 'Cubone',
 'Marowak',
 'Hitmonlee',
 'Hitmonchan',
 'Lickitung',
 'Koffing',
 'Weezing',
 'Chansey',
 'Kangaskhan',
 'Horsea',
 'Seadra',

In [10]:
len(pokemon_with_weakness_less_than_4)

102

There are 102 pokemons which have less than 4 weaknesses.

#### Get all Pokemons that have no multipliers at all

In [11]:
pokemon_with_no_multipliers = list(df[df["multipliers"].isna()]["name"])

In [12]:
pokemon_with_no_multipliers

['Venusaur',
 'Charizard',
 'Blastoise',
 'Butterfree',
 'Beedrill',
 'Pidgeot',
 'Raticate',
 'Fearow',
 'Arbok',
 'Raichu',
 'Sandslash',
 'Nidoqueen',
 'Nidoking',
 'Clefable',
 'Ninetales',
 'Wigglytuff',
 'Golbat',
 'Vileplume',
 'Parasect',
 'Venomoth',
 'Dugtrio',
 'Persian',
 'Golduck',
 'Primeape',
 'Arcanine',
 'Poliwrath',
 'Alakazam',
 'Machamp',
 'Victreebel',
 'Tentacruel',
 'Golem',
 'Rapidash',
 'Slowbro',
 'Magneton',
 "Farfetch'd",
 'Dodrio',
 'Dewgong',
 'Muk',
 'Cloyster',
 'Gengar',
 'Onix',
 'Hypno',
 'Kingler',
 'Electrode',
 'Exeggutor',
 'Marowak',
 'Hitmonlee',
 'Hitmonchan',
 'Lickitung',
 'Weezing',
 'Rhydon',
 'Chansey',
 'Tangela',
 'Kangaskhan',
 'Seadra',
 'Seaking',
 'Starmie',
 'Mr. Mime',
 'Scyther',
 'Jynx',
 'Electabuzz',
 'Magmar',
 'Pinsir',
 'Tauros',
 'Gyarados',
 'Lapras',
 'Ditto',
 'Vaporeon',
 'Jolteon',
 'Flareon',
 'Porygon',
 'Omastar',
 'Kabutops',
 'Aerodactyl',
 'Snorlax',
 'Articuno',
 'Zapdos',
 'Moltres',
 'Dragonite',
 'Mewtwo',
 '

In [13]:
len(pokemon_with_no_multipliers)

81

There are 81 pokemons which have less than 4 weaknesses.

#### Get all Pokemons that do not have more than 2 evolutions

In [14]:
# Check for conversion to embedded list using ast module

x = "[{'num': 1, 'name': 'xasdsad'}, {'num': 2, 'name': 'djad'}]"

import ast
x = ast.literal_eval(x)
len(x)

2

In [15]:
# Define a function for getting length of embedded list

def get_item_count(s) -> int:
    """
    Returns the count of items embedded within the string
    """

    
    # "literal_eval" evaluates an expression node or a string containing only a Python literal or container display. 
    # The string or node provided may only consist of the following Python literal structures: strings, 
    # bytes, numbers, tuples, lists, dicts, sets, booleans, None and Ellipsis.
    
    try:
        if isinstance(s, str):
            x = ast.literal_eval(s)
            return len(x)  
        else:
            return s
    except Exception as e:
        raise(e)

In [16]:
# Create a copy of df for analysis
df_has_next_evolution = df.copy()

# Replace Nan with 0
df_has_next_evolution["next_evolution"].fillna(0, inplace = True)

# Evaluating the problem
pokemon_with_no_more_than_2_evolutions = list(df_has_next_evolution[df_has_next_evolution["next_evolution"].
                                                                    apply(lambda x: get_item_count(x)) <= 2]["name"])

In [17]:
pokemon_with_no_more_than_2_evolutions

['Bulbasaur',
 'Ivysaur',
 'Venusaur',
 'Charmander',
 'Charmeleon',
 'Charizard',
 'Squirtle',
 'Wartortle',
 'Blastoise',
 'Caterpie',
 'Metapod',
 'Butterfree',
 'Weedle',
 'Kakuna',
 'Beedrill',
 'Pidgey',
 'Pidgeotto',
 'Pidgeot',
 'Rattata',
 'Raticate',
 'Spearow',
 'Fearow',
 'Ekans',
 'Arbok',
 'Pikachu',
 'Raichu',
 'Sandshrew',
 'Sandslash',
 'Nidoran ♀ (Female)',
 'Nidorina',
 'Nidoqueen',
 'Nidoran ♂ (Male)',
 'Nidorino',
 'Nidoking',
 'Clefairy',
 'Clefable',
 'Vulpix',
 'Ninetales',
 'Jigglypuff',
 'Wigglytuff',
 'Zubat',
 'Golbat',
 'Oddish',
 'Gloom',
 'Vileplume',
 'Paras',
 'Parasect',
 'Venonat',
 'Venomoth',
 'Diglett',
 'Dugtrio',
 'Meowth',
 'Persian',
 'Psyduck',
 'Golduck',
 'Mankey',
 'Primeape',
 'Growlithe',
 'Arcanine',
 'Poliwag',
 'Poliwhirl',
 'Poliwrath',
 'Abra',
 'Kadabra',
 'Alakazam',
 'Machop',
 'Machoke',
 'Machamp',
 'Bellsprout',
 'Weepinbell',
 'Victreebel',
 'Tentacool',
 'Tentacruel',
 'Geodude',
 'Graveler',
 'Golem',
 'Ponyta',
 'Rapidash',

In [18]:
len(pokemon_with_no_more_than_2_evolutions)

150

In [19]:
df_has_next_evolution[df_has_next_evolution["next_evolution"].apply(lambda x: get_item_count(x)) > 2]["name"]

132    Eevee
Name: name, dtype: object

There is only 1 pokemon which has more than 2 evolutions in this dataset. The name is 'Eevee'. Rest of all the pokemons have less than or equal to 2 pokemons

#### Get all Pokemons whose spawn time is less than 300 seconds.

In [20]:
# Checking if spawn_time has null values

df["spawn_time"].isna().sum()

6

In [21]:
# Define function to get time in seconds

def get_time_in_seconds(s) -> int:
    """
    Converts the time format from mm:ss to float(ss)
    """
    try:
        if isinstance(s, str):
            mins = float(s[:2])
            secs = float(s[-2:])

            time = mins*60 + secs

            return time
        else:
            return s
    except Exception as e:
        raise(e)

In [22]:
# Create a copy of df for analysis
df_spawn_time = df.copy()

# Replace Nan with 0
df_spawn_time["spawn_time"].fillna(0, inplace = True)

# Evaluating the problem statement
pokemon_with_spawn_time_less_than_300sec = list(df_spawn_time[(df_spawn_time["spawn_time"].
                                                                    apply(lambda x: get_time_in_seconds(x)) < 300)
                                                             &
                                                             (df_spawn_time["spawn_time"].
                                                                    apply(lambda x: get_time_in_seconds(x)) > 0)]["name"])

In [23]:
pokemon_with_spawn_time_less_than_300sec

['Squirtle',
 'Blastoise',
 'Metapod',
 'Weedle',
 'Kakuna',
 'Beedrill',
 'Pidgey',
 'Pidgeotto',
 'Pidgeot',
 'Rattata',
 'Raticate',
 'Fearow',
 'Arbok',
 'Pikachu',
 'Sandshrew',
 'Nidoran ♀ (Female)',
 'Nidoran ♂ (Male)',
 'Clefairy',
 'Clefable',
 'Ninetales',
 'Golbat',
 'Oddish',
 'Paras',
 'Parasect',
 'Venonat',
 'Diglett',
 'Meowth',
 'Persian',
 'Psyduck',
 'Growlithe',
 'Arcanine',
 'Poliwag',
 'Poliwrath',
 'Abra',
 'Machop',
 'Machamp',
 'Bellsprout',
 'Tentacool',
 'Graveler',
 'Ponyta',
 'Rapidash',
 'Slowbro',
 'Magnemite',
 "Farfetch'd",
 'Dodrio',
 'Muk',
 'Cloyster',
 'Gastly',
 'Haunter',
 'Gengar',
 'Onix',
 'Drowzee',
 'Hypno',
 'Krabby',
 'Kingler',
 'Voltorb',
 'Electrode',
 'Cubone',
 'Marowak',
 'Hitmonlee',
 'Lickitung',
 'Rhyhorn',
 'Chansey',
 'Kangaskhan',
 'Horsea',
 'Seadra',
 'Goldeen',
 'Mr. Mime',
 'Electabuzz',
 'Pinsir',
 'Tauros',
 'Gyarados',
 'Jolteon',
 'Porygon',
 'Kabuto']

In [24]:
len(pokemon_with_spawn_time_less_than_300sec)

75

There are 75 pokemons which have less than 300 sec spawn time excluding the ones which do not have any information regarding spawn time.

#### Get all Pokemon who have more than two types of capabilities

In [25]:
df.head(2)

id  num       name                                               img  \
0   1    1  Bulbasaur  http://www.serebii.net/pokemongo/pokemon/001.png   
1   2    2    Ivysaur  http://www.serebii.net/pokemongo/pokemon/002.png   

                  type  height  weight            candy  candy_count  \
0  ['Grass', 'Poison']    0.71     6.9  Bulbasaur Candy         25.0   
1  ['Grass', 'Poison']    0.99    13.0  Bulbasaur Candy        100.0   

           egg  spawn_chance  avg_spawns spawn_time multipliers  \
0            2         0.690        69.0      20:00      [1.58]   
1  Not in Eggs         0.042         4.2      07:00  [1.2, 1.6]   

                             weaknesses  \
0  ['Fire', 'Ice', 'Flying', 'Psychic']   
1  ['Fire', 'Ice', 'Flying', 'Psychic']   

                                      next_evolution  \
0  [{'num': '002', 'name': 'Ivysaur'}, {'num': '0...   
1               [{'num': '003', 'name': 'Venusaur'}]   

                          prev_evolution  
0                                    NaN  
1  [{'num': '001', 'name': 'Bulbasaur'}]

In [26]:
# Create a copy of df for analysis
df_has_capabilities = df.copy()

# Replace Nan with 0
df_has_capabilities["type"].fillna(0, inplace = True)

# Evaluating the problem statement
pokemon_has_more_than_2_capabilities = list(df_has_capabilities[df_has_capabilities["type"].
                                                                    apply(lambda x: get_item_count(x)) > 2]["name"])

In [27]:
pokemon_has_more_than_2_capabilities

[]

There is no pokemon which has more than 2 capabilities.